pkg/dispatcher/config.go

```diff
@@ -120,6 +154,38 @@ func (config *Config) DetermineClusterForJob(jobBase prowconfig.JobBase, path st
 		return config.SSHBastion, false, nil
 	}
 	if jobBase.Labels != nil {
+
+		requiredCapabilities := extractRequiredCapabilities(jobBase.Labels)
+		if len(requiredCapabilities) > 0 {
+			matchingClusters := []string{}
+			matchingClustersByProvider := map[string][]string{}
+
+			for clusterName, clusterInfo := range cm {
+				if matchesAllCapabilities(clusterInfo.Capabilities, requiredCapabilities) {
+					matchingClusters = append(matchingClusters, clusterName)
+					provider := clusterInfo.Provider
+					matchingClustersByProvider[provider] = append(matchingClustersByProvider[provider], clusterName)
+				}
+			}
+			if config.DetermineE2EByJob {
+				if cloud := config.DetermineCloudMapping(jobBase); cloud != "" {
+					if clusters, ok := matchingClustersByProvider[cloud]; ok {
+						if len(clusters) > 0 {
+							// as in other places in this file, use this method to have basic deterministic distribution
+							return api.Cluster(clusters[len(filepath.Base(path))%len(clusters)]), false, nil
```

Comment:
Can we extract the "basic deterministic distribution" logic to a function?

Reply:
We could but it would be very strange one liner as every time basically we are diving against different table len